pennylane/hf/tapering.py

```diff
@@ -0,0 +1,277 @@
+
+    **Example**
+
+    .. code-block::
+        >>> binary_matrix = np.array([[1, 0, 0, 0, 0, 1, 0, 0],
+                                      [1, 0, 1, 0, 0, 0, 1, 0],
+                                      [0, 0, 0, 1, 1, 0, 0, 1]])
+        >>> _reduced_row_echelon(binary_matrix)
+         array([[1, 0, 0, 0, 0, 1, 0, 0],
+                [0, 0, 1, 1, 1, 1, 1, 1],
+                [0, 0, 0, 1, 1, 0, 0, 1]])
+
+    """
+
+    rref_binary_matrix = deepcopy(binary_matrix)
+    shape = rref_binary_matrix.shape
+
+    for irow, icol in zip(range(shape[0]), range(shape[1])):
+
+        # find value and index of largest element in remainder of column icol
+        krow = irow + np.argmax(rref_binary_matrix[irow:, icol])
+
+        # swap rows krow and irow
+        rref_binary_matrix[krow], rref_binary_matrix[irow] = deepcopy(
+            rref_binary_matrix[irow]
+        ), deepcopy(rref_binary_matrix[krow])
+
+        # store remainder columns of the row irow
+        pvtcols = rref_binary_matrix[irow, icol:]
+
+        # get the column icol and set its irow element to 0 to avoid XORing pivot row with itself
+        currcol = deepcopy(rref_binary_matrix[:, icol])
+        currcol[irow] = 0
+        rref_binary_matrix[:, icol:] ^= np.outer(currcol, pvtcols)
+
+    return rref_binary_matrix.astype(int)
+
+
+def _kernel(binary_matrix):
+    r"""Computes the kernel of a binary matrix on the binary finite field :math:`\mathbb{Z}_2`.
+
+    Args:
+        binary_matrix (ndarray): binary matrix representation of the Hamiltonian.
+    Returns:
+        nullspace (ndarray): nullspace of the `binary_matrix` where each row correspond to a
+        basis vector in the nullspace.
+
+    **Example**
+
+    .. code-block::
+        >>> binary_matrix = np.array([[1, 0, 0, 0, 0, 1, 0, 0],
+                                      [0, 0, 1, 1, 1, 1, 1, 1],
+                                      [0, 0, 0, 1, 1, 0, 0, 1]])
+        >>> _kernel(binary_matrix)
+         array([[0, 1, 0, 0, 0, 0, 0, 0],
+                [0, 0, 1, 1, 1, 0, 0, 0],
+                [1, 0, 1, 0, 0, 1, 0, 0],
+                [0, 0, 1, 0, 0, 0, 1, 0],
+                [0, 0, 1, 1, 0, 0, 0, 1]])
+
+    """
+
+    # Get the columns with and without pivots
+    pivots = (binary_matrix.T != 0).argmax(axis=0)
+    nonpivots = np.setdiff1d(range(len(binary_matrix[0])), pivots)
+
+    # Initialize the nullspace
+    null_vector = np.zeros((binary_matrix.shape[1], len(nonpivots)), dtype=int)
+    null_vector[nonpivots, np.arange(len(nonpivots))] = 1
+
+    # Fill up the nullspace vectors from the binary matrix
+    null_vector_indices = np.ix_(pivots, np.arange(len(nonpivots)))
+    binary_vector_indices = np.ix_(np.arange(len(pivots)), nonpivots)
+    null_vector[null_vector_indices] = -binary_matrix[binary_vector_indices] % 2
+
+    nullspace = null_vector.T
+    return nullspace
+
+
+def generate_taus(nullspace, num_qubits):
```

Comment:
Maybe change the name to `get_generators`?

Reply:
Done!